pkg/appmanager/resourceConfig.go

```diff
@@ -1758,6 +2231,291 @@ func (rc *ResourceConfig) RemovePoolAt(offset int) bool {
+					for j := range deletedActionIndices {
+						deletedActionIndices[j]--
+					}
+				}
+
+				// Delete the merged rule if everything has been unmerged
+				if len(policy.Rules[i].Actions) == 0 {
+					policy.Rules = append(policy.Rules[:i], policy.Rules[i+1:]...)
+				}
+				break
+			}
+		}
+
+		// Delete the policy if its rules are empty
+		if len(policy.Rules) == 0 {
+			rc.RemovePolicy(*policy)
+		} else {
+			rc.SetPolicy(*policy)
+		}
+	}
+
+	// Delete entry from the mergedRulesMap if it is empty for this config
+	if len(mergedRulesMap[rsName]) == 0 {
+		delete(mergedRulesMap, rsName)
+	}
+	return true
+}
+
+func (rc *ResourceConfig) MergeRules(mergedRulesMap map[string]map[string]mergedRuleEntry) {
+	// Single service ingresses do not have rules that need merging
+	policy := rc.FindPolicy("forwarding")
+	rules := policy.Rules
+	if policy.Name != "url-rewrite-app-root-policy" {
+		var iDeletedRuleIndices []int
+		var jDeletedRuleIndices []int
+
+		// Iterate through the rules and compare them to each other
+		for i := range rules {
+			// Do not merge the same rule to itself or to rules that have already been merged
+			for j := i + 1; j < len(rules); j++ {
+				numMatches := 0
+				numIConditions := len(rules[i].Conditions)
+				numJConditions := len(rules[j].Conditions)
+				if numIConditions == numJConditions {
+					for k := range rules[i].Conditions {
+						for l := range rules[j].Conditions {
+							kConditionName := rules[i].Conditions[k].Name
+							lConditionName := rules[j].Conditions[l].Name
+							rules[i].Conditions[k].Name = ""
+							rules[j].Conditions[l].Name = ""
+							if reflect.DeepEqual(rules[i].Conditions[k], rules[j].Conditions[l]) {
+								numMatches++
+							}
+							rules[i].Conditions[k].Name = kConditionName
+							rules[j].Conditions[l].Name = lConditionName
+						}
+					}
+
+					// Only merge if both sets of conditions match
+					if numMatches == numIConditions {
+						var mergerEntry mergedRuleEntry
+						var mergeeEntry mergedRuleEntry
+
+						iName := rules[i].Name
+						jName := rules[j].Name
+						// Merge rule[i] into rule[j]
+						if ((strings.Contains(iName, "app-root") || strings.Contains(iName, "url-rewrite")) && !(strings.Contains(jName, "app-root") || strings.Contains(jName, "url-rewrite"))) ||
+							((strings.Contains(iName, "app-root") || strings.Contains(iName, "url-rewrite")) && (strings.Contains(jName, "app-root") || strings.Contains(jName, "url-rewrite"))) {
+							iDeletedRuleIndices = append(iDeletedRuleIndices, i)
+							mergerEntry.RuleName = jName
+							mergeeEntry.RuleName = iName
+							mergerEntry.OtherRuleNames = []string{iName}
+							mergeeEntry.OtherRuleNames = []string{jName}
+							mergerEntry.OriginalActions = rules[j].Actions
+							mergeeEntry.OriginalActions = rules[i].Actions
+
+							// Merge only unique actions
+							for k := range rules[i].Actions {
+								found := false
+								for l := range rules[j].Actions {
```

Comment:
See if this can be functionalized

Reply:
Tried doing this I am not coming up with anything quickly since it looks like rules are being passed by val rather than ref so due to time constraints I am going to keep this since it works.